File: packages/bots/src/baseline.ts
/**
 * The two measuring sticks.
 *
 * Neither is a difficulty level and neither should ever be offered to a player.
 * They exist so that a change to a real bot can be stated as a number: a bot
 * that cannot beat `greedy` by a wide margin has a bug, not a tuning problem.
 */

import { nextInt, type GameAction, type GameView, type HexId } from '@wc/shared';
import { pickLegal, type Bot, type BotContext } from './types.js';

/** Uniform over the legal actions. The floor. */
export const RandomBot: Bot = {
  name: 'random',
  chooseMove(view: GameView, ctx: BotContext): GameAction {
    const legal = pickLegal(view);
    return legal[nextInt(ctx.rng, legal.length)] as GameAction;
  },
};

/**
 * Take a location if one is on offer, otherwise hit something, preferring a hit
 * that destroys. No lookahead, no board sense — it plays the action in front of
 * it. Ties are broken with the context rng so a match is still reproducible.
 */
export const GreedyBot: Bot = {
  name: 'greedy',
  chooseMove(view: GameView, ctx: BotContext): GameAction {
    const legal = pickLegal(view);
    let best: GameAction[] = [];
    let bestScore = -Infinity;
    for (const action of legal) {
      const score = rank(view, action);
      if (score > bestScore) {
        bestScore = score;
        best = [action];
      } else if (score === bestScore) {
        best.push(action);
      }
    }
    return best[nextInt(ctx.rng, best.length)] as GameAction;
  },
};

function target(action: GameAction): HexId | undefined {
  if (action.type === 'attack' || action.type === 'followAttack') return action.to;
  if (action.type === 'tactic') return action.target;
  return undefined;
}

function rank(view: GameView, action: GameAction): number {
  // Claiming a location is the only thing that wins the game.
  if (action.type === 'control' || action.type === 'followControl') return 100;

  const hit = target(action);
  if (hit !== undefined) {
    const stack = view.units[hit];
    // A single-coin stack dies to this hit; a bolstered one only shrinks.
    if (stack && stack.team !== view.players[view.you]?.team) return stack.coins === 1 ? 90 : 70;
    return 60; // a fortification, or a tactic that hits an empty hex
  }

  switch (action.type) {
    case 'deploy':
      return 50;
    case 'move':
    case 'followMove':
    case 'tactic':
      return 40;
    case 'bolster':
      return 30;
    case 'recruit':
    case 'followRecruit':
      return 20;
    case 'draft':
    case 'ban':
      return 10;
    // Discarding a coin facedown does nothing but end the turn.
    case 'pass':
      return 0;
    default:
      return 15;
  }
}

export const BASELINE_BOTS: Readonly<Record<string, Bot>> = {
  random: RandomBot,
  greedy: GreedyBot,
};

File: crates/wc-bot/src/baseline.rs
//! The two measuring sticks.
//!
//! Neither is a difficulty level and neither should ever be offered to a player.
//! They exist so that a change to a real bot can be stated as a number: a bot
//! that cannot beat `greedy` by a wide margin has a bug, not a tuning problem.

use crate::position::Position;
use wc_core::board::{HexIdx, NONE};
use wc_core::rng::Rng;
use wc_core::types::{Action, ActionKind, Seat};

/// Uniform over the legal actions. The floor.
pub fn random_choice(legal: &[Action], rng: &mut Rng) -> Action {
    legal[rng.next_int(legal.len())]
}

/// Take a location if one is on offer, otherwise hit something, preferring a hit
/// that destroys. No lookahead, no board sense — it plays the action in front of
/// it. Ties are broken with the caller's rng, so a match is still reproducible.
pub fn greedy_choice<P: Position>(
    pos: &P,
    you: Seat,
    legal: &[Action],
    rng: &mut Rng,
    scratch: &mut Vec<u32>,
) -> Action {
    scratch.clear();
    let mut best_score = i32::MIN;
    for (i, action) in legal.iter().enumerate() {
        let score = rank(pos, you, *action);
        if score > best_score {
            best_score = score;
            scratch.clear();
            scratch.push(i as u32);
        } else if score == best_score {
            scratch.push(i as u32);
        }
    }
    legal[scratch[rng.next_int(scratch.len())] as usize]
}

fn target(action: Action) -> Option<HexIdx> {
    match action.kind {
        ActionKind::Attack | ActionKind::FollowAttack => Some(action.to),
        ActionKind::Tactic if action.target != NONE => Some(action.target),
        _ => None,
    }
}

fn rank<P: Position>(pos: &P, you: Seat, action: Action) -> i32 {
    use ActionKind::*;
    // Claiming a location is the only thing that wins the game.
    if matches!(action.kind, Control | FollowControl) {
        return 100;
    }

    if let Some(hit) = target(action) {
        let mine = pos.team_of(you);
        return match pos.units().get(hit) {
            // A single-coin stack dies to this hit; a bolstered one only shrinks.
            Some(stack) if stack.team != mine => {
                if stack.coins == 1 {
                    90
                } else {
                    70
                }
            }
            // A fortification, or a tactic that hits an empty hex.
            _ => 60,
        };
    }

    match action.kind {
        Deploy => 50,
        Move | FollowMove | Tactic => 40,
        Bolster => 30,
        Recruit | FollowRecruit => 20,
        Draft | Ban => 10,
        // Discarding a coin facedown does nothing but end the turn.
        Pass => 0,
        _ => 15,
    }
}
